ReadMe.md
bashplus
========

A Collection of Useful Bash Functions


## Synopsis

```
source "$PROJECT_ROOT/.bpan/lib/bpan.bash" --
bpan:source bashplus

+on-macos && echo "I'm on a Mac"
```


## Description

The bashplus library defines dozens of useful Bash functions to make Bash
programs much more readable.


## Functions

* `+assert-cmd <name>`
* `+assert-cmd-ver <name> <version>`
* `+assert-perl [<perl-version>]`
* `+assert-sed [<sed-version>]`
* `+can <name>`
* `die <msg>`
* `+is-bash32+`
* `+is-bash40+`
* `+is-bash41+`
* `+is-bash42+`
* `+is-bash43+`
* `+is-bash44+`
* `+is-bash50+`
* `+is-bash51+`
* `+is-cmd <name>`
* `+is-cmd-ver <name> <version>`
* `+is-file-same <file1> <file2>`
* `+is-file-diff <file1> <file2>`
* `+is-func <name>`
* `+is-online`
* `+mktemp`
* `+path`
* `+sort`
* `+sym [<prefix>]`
* `+trap <bash-code> [<signal>]`
* `warn <msg>`


## Copyright and License

Copyright 2022 by Ingy döt Net

This is free software, licensed under:

The MIT (X11) License


Changes
[version "0.1.69"]
date = 2023-01-01T12:28:24Z
done = Update .bpan/config

[version "0.1.68"]
date = 2022-12-30T17:26:19Z
done = Refactor 'set -x' code
done = Revert "Remove BASHPLUS_DEBUG_BASH_X support"

[version "0.1.67"]
date = 2022-12-21T18:58:12Z
done = Add +func:list support

[version "0.1.66"]
date = 2022-12-21T18:44:14Z
done = Revert "Make +time:stamp be UTC"

[version "0.1.65"]
date = 2022-12-21T18:12:16Z
done = Remove BASHPLUS_DEBUG_BASH_X support

[version "0.1.64"]
date = 2022-12-18T18:42:44Z
done = Update bpan assets
done = Make +time:stamp be UTC

[version "0.1.63"]
date = 2022-12-15T19:18:47Z
done = Add type to .bpan/config

[version "0.1.62"]
date = 2022-12-15T09:50:56Z
done = Fix +time:stamp to use iso8601 'z' format

[version "0.1.61"]
date = Sun Dec 11 05:38:56 PM PST 2022
done = Config changes

[version "0.1.60"]
date = Thu Dec  8 05:50:09 PM PST 2022
done = Change config to '[author "github:ingydotnet"]'

[version "0.1.59"]
date = Mon Dec  5 12:48:48 PM PST 2022
done = Fix config

[version "0.1.58"]
date = Mon Dec  5 10:49:25 AM PST 2022
done = Change package.repo to package.owner

[version "0.1.57"]
date = Sun Nov 20 12:35:35 PM PST 2022
done = Limit +source to BASHPLUS_PATH or local .bpan/lib

[version "0.1.56"]
date = Sat Nov 19 10:53:26 AM PST 2022
done = Add +str:repeat
done = Move Meta file to .bpan/config

[version "0.1.55"]
date = Sat Nov 19 09:07:12 AM PST 2022
done = Add str.bash

[version "0.1.54"]
date = Thu Nov 17 09:58:40 PM PST 2022
done = Add tag to Meta

[version "0.1.53"]
date = Thu Nov 17 09:56:16 PM PST 2022
done = New BPAN config format

[version "0.1.52"]
date = Thu Nov 17 09:46:18 PM PST 2022
done = Allow optional 2nd arg 'remote' for +git:tag-pushed

[version "0.1.51"]
date = Fri Nov 11 10:07:03 AM PST 2022
done = Change source bashplus --foo to +foo

[version "0.1.50"]
date = Fri Nov 11 06:46:32 AM PST 2022
done = Avoid unnecessary subshell functions

[version "0.1.49"]
date = Fri Nov 11 06:18:11 AM PST 2022
done = Finish fun -> func refactor

[version "0.1.48"]
date = Fri Nov 11 05:52:08 AM PST 2022
done = Add a +time:stamp function
done = Rename bashplus/fun to bashplus/func
done = Add ::FUNC-NAME:: support to bashplus/fun

[version "0.1.47"]
date = Wed Nov  9 10:03:54 AM PST 2022
done = Add +time:ymd for yyyy-mm-dd date string
done = Syntax tweak

[version "0.1.46"]
date = Sun Nov  6 11:33:43 AM PST 2022
done = Support BASHPLUS_DEBUG_BASH_X

[version "0.1.45"]
date = Wed Nov  2 03:01:23 PM PDT 2022
done = Implement +stack-trace and use in die and error

[version "0.1.44"]
date = Wed Nov  2 11:04:10 AM PDT 2022
done = Simplfy +source

[version "0.1.43"]
date = Sun Oct 30 01:25:20 PM PDT 2022
done = Fix a wording bug in the 'error' function

[version "0.1.42"]
date = Sun Oct 30 01:13:23 PM PDT 2022
done = Fix bugs in 'error' function from refactor

[version "0.1.41"]
date = Sun Oct 30 12:56:01 PM PDT 2022
done = A couple more refactoring bugs fixed

[version "0.1.40"]
date = Sun Oct 30 11:48:55 AM PDT 2022
done = Another refactoring bug fix

[version "0.1.39"]
date = Sun Oct 30 11:32:31 AM PDT 2022
done = Small fixes

[version "0.1.38"]
date = Sun Oct 30 10:58:51 AM PDT 2022
done = Fixes and refactoring

[version "0.1.37"]
date = Sun Oct 30 09:51:25 AM PDT 2022
done = Break up bashplus into smaller sub libraries
done = Add +say in lib/bashplus/say.bash

[version "0.1.36"]
date = Sat Oct 29 09:21:42 PM PDT 2022
done = Add sub-library bashplus/git.bash

[version "0.1.35"]
date = Sat Oct 29 03:48:09 PM PDT 2022
done = Renumber tests

[version "0.1.34"]
date = Sat Oct 29 03:47:29 PM PDT 2022
done = Fix some bugs in +timer-wrap

[version "0.1.33"]
date = Sat Oct 29 01:35:47 PM PDT 2022
done = Improve +timer* functions

[version "0.1.32"]
date = Sat Oct 29 12:02:46 PM PDT 2022
done = Add timer functions
done = Add +source and +wrap functions

[version "0.1.31"]
date = Mon Oct 17 09:11:17 PM PDT 2022
done = Reformat bashplus:version function

[version "0.1.30"]
date = Sat Oct 15 05:59:01 PM PDT 2022
done = Fix whether to print line # for die message

[version "0.1.29"]
date = Thu Oct 13 09:16:41 AM PDT 2022
done = Newline one last die msg line disables line #

[version "0.1.28"]
date = Wed Oct  5 07:58:51 PM PDT 2022
done = Fix .bpan/config symlink

[version "0.1.27"]
date = Wed Oct  5 09:39:31 AM PDT 2022
done = Fix +mtime on macos

[version "0.1.26"]
date = Sun Oct  2 06:16:04 PM PDT 2022


[version "0.1.25"]
date = Sun Oct  2 06:14:47 PM PDT 2022
done = Add +mtime and +time

[version "0.1.24"]
date = Sat Oct  1 02:04:37 PM PDT 2022
done = Add +dirname and +dirpath
done = Version 0.1.23

[version "0.1.23"]
date = Fri Sep 30 12:56:05 PM PDT 2022
done = Change 'update.template' to 'update.file' in config

[version "0.1.22"]
date = Fri Sep 30 09:22:40 AM PDT 2022
done = Don't use 'warn' internally

[version "0.1.21"]
date = Fri Sep 30 08:26:51 AM PDT 2022
done = Add +is-empty function for files and dirs
done = Change 'function x ...' to 'x() ...'
done = Change '+warn' back to 'warn'

[version "0.1.20"]
date = Mon Sep 26 07:55:18 AM EDT 2022
done = Fix bug in test

[version "0.1.19"]
date = Mon Sep 26 07:53:05 AM EDT 2022
done = Add more possible +sym implementations
done = Remove .bpan/lib/bpan.bash

[version "0.1.18"]
date = Sun Sep 25 12:34:56 PM EDT 2022
done = BPAN update

[version "0.1.17"]
date = Sat Sep 24 16:23:18 EDT 2022
done = Dummy commit to test release

[version "0.1.16"]
date = Sat Sep 24 16:20:52 EDT 2022
done = Dummy commit to test release

[version "0.1.15"]
date = Sat Sep 24 16:13:12 EDT 2022
done = Dummy commit to test releease

[version "0.1.14"]
date = Sat Sep 24 16:01:14 EDT 2022
done = Dummy commit for testing release

[version "0.1.13"]
date = Sat Sep 24 03:18:20 PM EDT 2022
done = Removed a blank line for testing release

[version "0.1.12"]
date = Sat Sep 24 03:06:12 PM EDT 2022
done = Added a blank line

[version "0.1.11"]
date = Sat Sep 24 02:14:24 PM EDT 2022
done = Add comments to library

[version "0.1.10"]

lib/bashplus/say.bash
# Normal color vars:
D='\e[30m'      # black (dark)
R='\e[31m'      # red
G='\e[32m'      # green
Y='\e[33m'      # yellow
B='\e[34m'      # blue
M='\e[35m'      # magenta
C='\e[36m'      # cyan
W='\e[37m'      # white

# Bold color vars:
DD='\e[30;1m'   # black (dark)
RR='\e[31;1m'   # red
GG='\e[32;1m'   # green
YY='\e[33;1m'   # yellow
BB='\e[34;1m'   # blue
MM='\e[35;1m'   # magenta
CC='\e[36;1m'   # cyan
WW='\e[37;1m'   # white

# Reset color var:
Z='\e[0m'       # reset

say() (
  +say "$@"
)

+say() (
  ${option_quiet:-false} && return

  if [[ ${1-} == -? ]]; then
    color=${1#-}; shift
    "+say:$color" "$@"
  else
    +say:color "" "$@"
  fi
)

+say:color() (
  ${option_quiet:-false} && return

  color=$1; shift
  if [[ -t 1 ]]; then
    echo -e "$Z$color$*$Z"
  else
    echo -e "$*"
  fi
)

+say:d() { +say:color "$D" "$@"; }
+say:r() { +say:color "$R" "$@"; }
+say:g() { +say:color "$G" "$@"; }
+say:y() { +say:color "$Y" "$@"; }
+say:b() { +say:color "$B" "$@"; }
+say:m() { +say:color "$M" "$@"; }
+say:c() { +say:color "$C" "$@"; }
+say:w() { +say:color "$W" "$@"; }

+say:D() { +say:color "$DD" "$@"; }
+say:R() { +say:color "$RR" "$@"; }
+say:G() { +say:color "$GG" "$@"; }
+say:Y() { +say:color "$YY" "$@"; }
+say:B() { +say:color "$BB" "$@"; }
+say:M() { +say:color "$MM" "$@"; }
+say:C() { +say:color "$CC" "$@"; }
+say:W() { +say:color "$WW" "$@"; }
